### PR TITLE
Cleanup low-level UCX options

### PR DESCRIPTION
**Pull request by madsbk**
Minimize the use of pure C

This PR is part **2** of an effort to merge https://github.com/rapidsai/ucx-py/pull/648 in a divide-and-conquer approach.

### CODE REVIEW

File: ucp/_libs/src/c_util.c [hunk 1]
```diff
@@ -11,79 +11,21 @@
...
-    param->err_mode           = err_cb == NULL ? UCP_ERR_HANDLING_MODE_NONE : UCP_ERR_HANDLING_MODE_PEER;
-    param->err_handler.cb     = err_cb;
-    param->err_handler.arg    = NULL;
-    param->conn_request       = conn_request;
-    return 0;
-}
 
-void c_util_get_ucp_ep_params_free(ucp_ep_params_t *param) {
-    free((void*) param->sockaddr.addr);
+void c_util_set_sockaddr_free(ucs_sock_addr_t *sockaddr) {
```
Comment (pentschev) [inline ucp/_libs/src/c_util.c:29]:
```suggestion
void c_util_sockaddr_free(ucs_sock_addr_t *sockaddr) {
```

Comment (pentschev) [reply ucp/_libs/src/c_util.c:29]: The name above seems a bit odd to me, suggesting a change but feel free to ignore it if you disagree.

File: ucp/_libs/src/c_util.h [hunk 2]
```diff
@@ -7,21 +7,6 @@
...
-                             const char *ip_address,
-                             uint16_t port,
-                             ucp_err_handler_cb_t err_cb);
-
-int c_util_get_ucp_ep_conn_params(ucp_ep_params_t *param,
-                                  ucp_conn_request_h conn_request,
-                                  ucp_err_handler_cb_t err_cb);
-
-void c_util_get_ucp_ep_params_free(ucp_ep_params_t *param);
+void c_util_set_sockaddr_free(ucs_sock_addr_t *sockaddr);
```
Comment (pentschev) [inline ucp/_libs/src/c_util.h:12]:
```suggestion
void c_util_sockaddr_free(ucs_sock_addr_t *sockaddr);
```

Comment (pentschev) [reply ucp/_libs/src/c_util.h:12]: Ditto.

File: ucp/_libs/ucx_api.pyx [hunk 3]
```diff
@@ -396,14 +398,24 @@ cdef class UCXWorker(UCXObject):
...
+            params.err_mode = UCP_ERR_HANDLING_MODE_PEER
+        params.err_handler.cb = err_cb
+        params.err_handler.arg = NULL
+        if c_util_set_sockaddr(&params.sockaddr, ip_address.encode(), port):
+            raise MemoryError("Failed allocation of sockaddr")
 
         cdef ucp_ep_h ucp_ep
         cdef ucs_status_t status = ucp_ep_create(self._handle, &params, &ucp_ep)
-        c_util_get_ucp_ep_params_free(&params)
+        c_util_set_sockaddr_free(&params.sockaddr)
```
Comment (pentschev) [inline ucp/_libs/ucx_api.pyx:418]:
```suggestion
        c_util_sockaddr_free(&params.sockaddr)
```

File: ucp/_libs/ucx_api_dep.pxd [hunk 4]
```diff
@@ -14,40 +14,15 @@ from libc.string cimport memset
...
 
-    int c_util_get_ucp_ep_params(ucp_ep_params_t *param,
-                                 const char *ip_address,
-                                 uint16_t port,
-                                 ucp_err_handler_cb_t err_cb)
-    int c_util_get_ucp_ep_conn_params(ucp_ep_params_t *param,
-                                      ucp_conn_request_h conn_request,
-                                      ucp_err_handler_cb_t err_cb)
-    void c_util_get_ucp_ep_params_free(ucp_ep_params_t *param)
+    void c_util_set_sockaddr_free(ucs_sock_addr_t *sockaddr)
```
Comment (pentschev) [inline ucp/_libs/ucx_api_dep.pxd:25]:
```suggestion
    void c_util_sockaddr_free(ucs_sock_addr_t *sockaddr)
```

File: ucp/_libs/ucx_api.pyx [hunk 5]
```diff
@@ -212,7 +212,11 @@ cdef class UCXContext(UCXObject):
         dict _config
         readonly bint cuda_support
 
-    def __init__(self, config_dict):
+    def __init__(
+        self,
+        config_dict,
+        future_flags=(UCP_FEATURE_TAG | UCP_FEATURE_WAKEUP | UCP_FEATURE_STREAM)
```
Comment (pentschev) [inline ucp/_libs/ucx_api.pyx:218]: Did you mean to name this `feature(s)_flags` ?

File: ucp/_libs/ucx_api.pyx [hunk 6]
```diff
@@ -656,16 +676,18 @@ cdef class UCXListener(UCXObject):
...
+        params.conn_handler.cb = _listener_cb
+        params.conn_handler.arg = <void*> self.cb_data
+        if c_util_set_sockaddr(&params.sockaddr, NULL, port):
+            raise MemoryError("Failed allocation of sockaddr")
 
         cdef ucs_status_t status = ucp_listener_create(
             worker._handle, &params, &self._handle
         )
-        c_util_get_ucp_listener_params_free(&params)
+        c_util_set_sockaddr_free(&params.sockaddr)
```
Comment (pentschev) [inline ucp/_libs/ucx_api.pyx:690]:
```suggestion
        c_util_sockaddr_free(&params.sockaddr)
```

**Review by pentschev — COMMENTED** [review]
Overall this LGTM, thanks @madsbk ! I have a couple nitpicks, let me know what you think.

File: ucp/_libs/ucx_api.pyx (hunk 5 above)
Comment (madsbk) [reply ucp/_libs/ucx_api.pyx]: Yes :)

**Comment by madsbk** [comment]
> Overall this LGTM, thanks @madsbk ! I have a couple nitpicks, let me know what you think.

Thanks @pentschev, I agree with all of it :)

**Review by pentschev — APPROVED** [review]
Thanks @madsbk !

**Pull request merged**